pyuvdata/uvdata/uvh5.py

```diff
@@ -136,6 +131,224 @@ def _write_complex_astype(data, dset, indices):
+
+    Parameters
+    ----------
+    s : slice object
+        Slice object to index with
+    axlen : int
+        Length of axis s slices into
+
+    Returns
+    -------
+    int
+        Length of slice object
+    """
+    if s.start is None:
+        start = 0
+    else:
+        start = s.start
+    if s.stop is None:
+        stop = axlen
+    else:
+        stop = np.min([s.stop, axlen])
+    if s.step is None:
+        step = 1
+    else:
+        step = s.step
+
+    return (stop - 1 - start) // step + 1
+
+
+def _get_dset_shape(dset, indices):
+    """
+    Given a 4-tuple of indices, determine the indexed array shape.
+
+    Parameters
+    ----------
+    dset : h5py dataset
+        A reference to an HDF5 dataset on disk.
+    indices : tuple
+        A 4-tuple with the indices to extract along each dimension of dset.
+        Each element should contain a list of indices, a slice element,
+        or a list of slice elements that will be concatenated after slicing.
+
+    Returns
+    -------
+    tuple
+        4-tuple with the shape of the indexed array
+    """
+    dset_shape = list(dset.shape)
+    for i, inds in enumerate(indices):
+        # check for integer
+        if isinstance(inds, (int, np.integer)):
+            dset_shape[i] = 1
+        # check for slice object
+        if isinstance(inds, slice):
+            dset_shape[i] = _get_slice_len(inds, dset_shape[i])
+        # check for list
+        if isinstance(inds, list):
+            # check for list of integers
+            if isinstance(inds[0], (int, np.integer)):
+                dset_shape[i] = len(inds)
+            elif isinstance(inds[0], slice):
+                dset_shape[i] = sum((_get_slice_len(s, dset_shape[i]) for s in inds))
+
+    return dset_shape
+
+
+def _index_dset(dset, indices):
+    """
+    Index a UVH5 data, flags or nsamples h5py dataset.
+
+    Parameters
+    ----------
+    dset : h5py dataset
+        A reference to an HDF5 dataset on disk.
+    indices : tuple
+        A 4-tuple with the indices to extract along each dimension of dset.
+        Each element should contain a list of indices, a slice element,
+        or a list of slice elements that will be concatenated after slicing.
+        Indices must be provided such that all dimensions can be indexed
+        simultaneously.
+
+    Returns
+    -------
+    ndarray
+        The indexed dset
+
+    Notes
+    -----
+    This makes and fills an empty array with dset indices.
+    For trivial indexing, (e.g. a trivial slice), constructing
+    a new array and filling it is suboptimal over direct
+    indexing, e.g. dset[indices].
+    This function specializes in repeated slices over the same axis,
+    e.g. if indices is [[slice(0, 5), slice(10, 15), ...], ..., ]
+    """
+    # get dset and arr shape
+    dset_shape = dset.shape
+    arr_shape = _get_dset_shape(dset, indices)
+
+    # create empty array of dset dtype
+    arr = np.empty(arr_shape, dtype=dset.dtype)
```

Comment:
I don't think you necessarily always want `dtype` of the array to be `dset.dtype`. For instance, with raw HERA correlator data, `dset.dtype` is a custom-defined type, whereas the output array should be `np.complex128` or `np.complex64`. We might just need to have special handling for this case, along the lines of what we do now.

Reply:
Yeah as you can see I haven't changed the `_read_complex_astype` for this reason. I guess perhaps keep this as is until we have a need for making it more flexible? Lmk what you think.

Reply:
Yes, but it looks like `_read_complex_astype` calls this internally now. Thinking about it a little bit more, my guess is that when the assignment happens (e.g., when we do `output_array.real = _index_dset(dset["r"], indices)`) the array is initially an integer type for HERA-correlator raw data because that's the dataset dtype, which then gets coerced into the declared type (`np.real32` or `np.real64`, depending on whether the output type is `np.complex64` or  `np.complex128`). I don't think it's that much of a performance penalty, since we'd have to convert the type eventually. It's just less explicit now, so maybe we can add a comment that the type conversion will happen outside of this function.

Reply:
Yes that's right--but I think that this is the same as it was before: `_read_complex_astype` would just assign `output_array.real` with `dset[...]`, which if it was an integer type would then be upcast into whatever dtype `output_array` was specified with. I can add a comment nonetheless. Where did you want this? In `_index_dset` or in `_read_complex_astype`?

Reply:
Yeah, maybe here where we make the placeholder output array, we just mention that it will natively be the same type as the data on disk, but may be upcast/type converted upon assignment.

Reply:
Okay but just to be clear, the function `_index_dset` is simply meant to index the dset with the same dtype as its written. The specialty function `_read_complex_astype` is meant specifically to upcast whats index by `_index_dset` into complex type. So I were to put a comment it would be in `_read_complex_astype`. I think the function is fairly self explanatory as is, but I'll add a comment anyways for good measure :)